src/Data/Constraint/Deriving/DeriveAll.hs

```diff
@@ -754,8 +757,11 @@ mtmiToExpression MatchingType {..} mi = do
   let extraTheta
             = filter (\t -> not $ any (eqType t . fst) bndrs) mtTheta
       tRepl = replaceTypeOccurrences mtBaseType mtNewType tOrig
-      tFun  = mkFunTys (extraTheta ++ map fst bndrs) tRepl
-      tvs   = tyCoVarsOfTypeWellScoped tFun
+      tFun  = mkVisFunTys (extraTheta ++ map fst bndrs) tRepl
+      tvs   =  tyCoVarsOfTypeWellScoped tFun
+#if __GLASGOW_HASKELL__ < 810
+      mkVisFunTys = mkFunTys
+#endif
```

Comment:
Probably, we also should use `mkInvisFunTys` here, unless there are any non-constraint types in `bndrs` (in which case we would need to split the list in two and apply  `mkVisFunTys` on the right side (but most likely these are also constraint-types only)).